// crisp-api/dist/resources/PluginConnect.js
"use strict";
/*
 * node-crisp-api
 *
 * Copyright 2022, Crisp IM SAS
 * Author: Baptiste Jamin <baptiste@crisp.chat>
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/**************************************************************************
 * IMPORTS
 ***************************************************************************/
// PROJECT: RESOURCES
const BaseResource_1 = __importDefault(require("./BaseResource"));
/**************************************************************************
 * CLASSES
 ***************************************************************************/
/**
 * Crisp PluginConnect Resource
 */
class PluginConnect extends BaseResource_1.default {
    /**
     * Get Connect Account
     */
    getConnectAccount() {
        return this.crisp.get(this.crisp.prepareRestUrl(["plugin", "connect", "account"]), {});
    }
    ;
    /**
     * Check Connect Session Validity
     */
    checkConnectSessionValidity() {
        return this.crisp.head(this.crisp.prepareRestUrl(["plugin", "connect", "session"]), {});
    }
    ;
    /**
     * List All Connect Websites
     */
    listAllConnectWebsites(pageNumber, filterConfigured) {
        // Generate query
        let query = {
            filter_configured: ((filterConfigured === true) ? "1" : "0")
        };
        return this.crisp.get(this.crisp.prepareRestUrl([
            "plugin", "connect", "websites", "all", String(pageNumber)
        ]), query);
    }
    ;
    /**
     * List Connect Websites Since
     */
    listConnectWebsitesSince(dateSince, filterConfigured) {
        // Generate query
        let query = {
            filter_configured: ((filterConfigured === true) ? "1" : "0")
        };
        if (dateSince) {
            // @ts-ignore
            query.date_since = dateSince;
        }
        return this.crisp.get(this.crisp.prepareRestUrl([
            "plugin", "connect", "websites", "since"
        ]), query);
    }
    ;
    /**
     * Get Connect Endpoints
     */
    getConnectEndpoints() {
        return this.crisp.get(this.crisp.prepareRestUrl(["plugin", "connect", "endpoints"]), {});
    }
    ;
}
/**************************************************************************
 * EXPORTS
 ***************************************************************************/
exports.default = PluginConnect;
