// crisp-api/dist/resources/WebsiteBase.js
"use strict";
/*
 * node-crisp-api
 *
 * Copyright 2022, Crisp IM SAS
 * Author: Baptiste Jamin <baptiste@crisp.chat>
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/**************************************************************************
 * IMPORTS
 ***************************************************************************/
// PROJECT: RESOURCES
const BaseResource_1 = __importDefault(require("./BaseResource"));
/**************************************************************************
 * CLASSES
 ***************************************************************************/
/**
 * Crisp WebsiteBase Resource
 */
class WebsiteBase extends BaseResource_1.default {
    /**
     * Check If Website Exists
     * @memberof WebsiteBase
     * @public
     * @method checkWebsiteExists
     * @param {string} domain
     * @return {Promise}
     */
    checkWebsiteExists(domain) {
        return this.crisp.head(this.crisp.prepareRestUrl(["website"]), {
            domain: domain
        });
    }
    ;
    /**
     * Create Website
     */
    createWebsite(websiteData) {
        return this.crisp.post(this.crisp.prepareRestUrl(["website"]), null, websiteData);
    }
    ;
    /**
     * Get A Website
     * @memberof WebsiteBase
     * @public
     * @method getWebsite
     * @param {string} websiteID
     * @return {Promise}
     */
    getWebsite(websiteID) {
        return this.crisp.get(this.crisp.prepareRestUrl(["website", websiteID]));
    }
    ;
    /**
     * Delete A Website
     */
    deleteWebsite(websiteID, verify) {
        return this.crisp.delete(this.crisp.prepareRestUrl(["website", websiteID]), undefined, {
            verify: verify
        });
    }
    ;
    /**
     * Abort Website Deletion
     */
    abortWebsiteDeletion(websiteID) {
        return this.crisp.delete(this.crisp.prepareRestUrl(["website", websiteID, "expunge"]));
    }
    ;
}
/**************************************************************************
 * EXPORTS
 ***************************************************************************/
exports.default = WebsiteBase;
