// crisp-api/dist/resources/WebsiteVisitors.js
"use strict";
/*
 * node-crisp-api
 *
 * Copyright 2022, Crisp IM SAS
 * Author: Valerian Saliou <valerian@valeriansaliou.name>
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/**************************************************************************
 * IMPORTS
 ***************************************************************************/
// PROJECT: RESOURCES
const BaseResource_1 = __importDefault(require("./BaseResource"));
/**************************************************************************
 * CLASSES
 ***************************************************************************/
/**
 * Crisp WebsiteVisitors Resource
 */
class WebsiteVisitors extends BaseResource_1.default {
    /**
     * Count Visitors
     */
    countVisitors(websiteID) {
        return this.crisp.get(this.crisp.prepareRestUrl(["website", websiteID, "visitors", "count"]));
    }
    ;
    /**
     * List Visitors
     */
    listVisitors(websiteID, pageNumber = 1) {
        return this.crisp.get(this.crisp.prepareRestUrl([
            "website", websiteID, "visitors", "list", String(pageNumber)
        ]));
    }
    ;
    /**
     * Pinpoint Visitors On A Map
     */
    pinpointVisitorsOnMap(websiteID, centerLongitude, centerLatitude, centerRadius) {
        // Generate query
        const query = {};
        if (typeof centerLongitude === "number") {
            query.center_longitude = String(centerLongitude);
        }
        if (typeof centerLatitude === "number") {
            query.center_latitude = String(centerLatitude);
        }
        if (typeof centerRadius === "number") {
            query.center_radius = String(centerRadius);
        }
        return this.crisp.get(this.crisp.prepareRestUrl([
            "website", websiteID, "visitors", "map"
        ]), query);
    }
    ;
    /**
     * Get Session Identifier From Token
     */
    getSessionIdentifierFromToken(websiteID, tokenID) {
        return this.crisp.get(this.crisp.prepareRestUrl([
            "website", websiteID, "visitors", "token", tokenID
        ]));
    }
    ;
    /**
     * Count Blocked Visitors
     */
    countBlockedVisitors(websiteID) {
        return this.crisp.get(this.crisp.prepareRestUrl(["website", websiteID, "visitors", "blocked"]));
    }
    ;
    /**
     * Count Blocked Visitors In Rule
     */
    countBlockedVisitorsInRule(websiteID, rule) {
        return this.crisp.get(this.crisp.prepareRestUrl([
            "website", websiteID, "visitors", "blocked", rule
        ]));
    }
    ;
    /**
     * Clear Blocked Visitors In Rule
     */
    clearBlockedVisitorsInRule(websiteID, rule) {
        return this.crisp.delete(this.crisp.prepareRestUrl([
            "website", websiteID, "visitors", "blocked", rule
        ]));
    }
    ;
}
/**************************************************************************
 * EXPORTS
 ***************************************************************************/
exports.default = WebsiteVisitors;
